### ISSUE TITLE
Callback is not called with local minima of initial guess in basinhopping algorithm

### ISSUE THREAD

**Issue by prince2110** [scipy/scipy#13011]
My issue is about ...

As far as I understand, callback function is called when a local minima is found. When I use basinhopping global minimization algorithm, I have observed that the callback function is not called when local minima is found for initial guess. I have tried to reproduce this issue with the following example:

```
from scipy.optimize import basinhopping
import numpy as np

def func(x):
    print("Objective function count: " + str(func.counter))
    func.counter += 1
    return np.random.rand(1)*1000

def callback_func(x, f, accepted):
    print("Callback count: " + str(callback_func.counter))
    callback_func.counter += 1

func.counter = 0
callback_func.counter = 0    
x0 = 2000

minimizer_kwargs = {"method": "COBYLA", "options":{"maxiter": 1}}
optimizer = basinhopping(func, x0, minimizer_kwargs=minimizer_kwargs, niter=2, callback=callback_func)
```

It produces following result:

```
Objective function count: 0
Objective function count: 1
Callback count: 0
Objective function count: 2
Callback count: 1
```
For the above mentioned example, both callback function and the objective function should have been called thrice but callback is called only twice.

#### Scipy/Numpy/Python version information:
 1.2.1 1.16.4 sys.version_info(major=3, minor=7, micro=4, releaselevel='final', serial=0)

**Comment by andyfaff** [comment]
The callback is [called once per basinhopping iteration](https://github.com/scipy/scipy/blob/master/scipy/optimize/_basinhopping.py#L688). You set `niter` to 2 for basinhopping, so I would only ever expect the callback to be called twice.

**Comment by prince2110** [comment]
With all due respect, I disagree with your previous comment. There is a local minima that is obtained from the initial guess and that information is not passed to the callback function. By this I mean to say that, suppose I am gathering all the predicted local minima's achieved by exploiting callback function then I will not be having any data regarding the local minima obtained by starting from initial guess. I will surely have data for further iterations but not for initial guess and this is where my problem is. 

I hope I am able to explain my problem properly.

**Comment by rgommers** [comment]
> There is a local minima that is obtained from the initial guess and that information is not passed to the callback function. By this I mean to say that, suppose I am gathering all the predicted local minima's achieved by exploiting callback function then I will not be having any data regarding the local minima obtained by starting from initial guess.

I'm not sure I understand what you mean. The initial guess is just that, a guess - it's not a minimum.

The code structure is simple here:
```
    for i in range(niter):
        if callable(callback):
            val = callback(...)
```
So @andyfaff's answer seems right, `niter` is an upper bound on the number of times the callback function is called.

**Comment by prince2110** [comment]
Maybe with the following example, I would be able to convey myself:

```
from scipy.optimize import basinhopping
import numpy as np

def func(x):
    print("Data passed to objective: " + str(x))
    print("Objective function count: " + str(func.counter))
    func.counter += 1
    return np.random.rand(1)*1000

def callback_func(x, f, accepted):
    print("Data passed to callback: " + str(x))
    print("Callback count: " + str(callback_func.counter))
    callback_func.counter += 1

func.counter = 0
callback_func.counter = 0    
x0 = 2000

np.random.seed(1)

minimizer_kwargs = {"method": "COBYLA", "options":{"maxiter": 5}}
optimizer = basinhopping(func, x0, minimizer_kwargs=minimizer_kwargs, niter=1, callback=callback_func)
print(optimizer)
```

The output is as follow:
```
Data passed to objective: [ 2000.]                                                                                                                                               
Objective function count: 0                                                                                                                                                      
Data passed to objective: [ 2001.]                                                                                                                                               
Objective function count: 1                                                                                                                                                      
Data passed to objective: [ 1999.]                                                                                                                                               
Objective function count: 2                                                                                                                                                      
Data passed to objective: [ 1998.]                                                                                                                                               
Objective function count: 3                                                                                                                                                      
Data passed to objective: [ 1998.5]                                                                                                                                              
Objective function count: 4                                                                                                                                                      
Data passed to objective: [ 1998.59233859]                                                                                                                                       
Objective function count: 5                                                                                                                                                      
Data passed to objective: [ 1999.59233859]                                                                                                                                       
Objective function count: 6                                                                                                                                                      
Data passed to objective: [ 1997.59233859]                                                                                                                                       
Objective function count: 7                                                                                                                                                      
Data passed to objective: [ 1998.09233859]                                                                                                                                       
Objective function count: 8                                                                                                                                                      
Data passed to objective: [ 1998.84233859]                                                                                                                                       
Objective function count: 9                                                                                                                                                      
Data passed to callback: 1998.5923385947688                                                                                                                                      
Callback count: 0                                                                                                                                                                
     nit: 1                                                                                                                                                                      
     fun: 0.11437481734488664                                                                                                                                                    
    nfev: 10                                                                                                                                                                     
 message: ['requested number of basinhopping iterations completed successfully']                                                                                                 
       x: array(1999.0)
```

For this example, I have used maxiter for COBYLA as 5 and the objective function was called 10 times for niter of 1. This shows that local minimization was performed twice but the data related to local minima was passed to callback only once.

**Comment by andyfaff** [comment]
@rgommers, @prince2110, I think I've got to the bottom of this.

- the callback function is indeed called once per [basinhopping iteration](https://github.com/scipy/scipy/blob/master/scipy/optimize/_basinhopping.py#L688).
- However, when the BasinHoppingRunner is [constructed](https://github.com/scipy/scipy/blob/master/scipy/optimize/_basinhopping.py#L678), there is an [initial minimisation](https://github.com/scipy/scipy/blob/master/scipy/optimize/_basinhopping.py#L72). This initial minimisation is not set to the callback function. As the documentation is written it probably should be, but that initial minimisation would not be after a basinhopping step had been taken. At the worst this would result in `niter+1` callbacks for `niter` steps.

**Comment by prince2110** [comment]
If my previous commit in the wrong PR solves this issue, then can I again submit a new pull request ?

**Comment by andyfaff** [comment]
I've already done one, #13029.
